src/nspawn/nspawn-mount.c

```diff
@@ -628,11 +688,107 @@ static int mount_legacy_cgroup_hierarchy(const char *dest, const char *controlle
         return 1;
 }
 
-static int mount_legacy_cgroups(
-                const char *dest,
+/* Mount a legacy cgroup hierarchy when cgroup namespaces are supported. */
+static int mount_legacy_cgns_supported(
                 bool userns, uid_t uid_shift, uid_t uid_range,
                 const char *selinux_apifs_context) {
+        _cleanup_set_free_free_ Set *controllers = NULL;
+        const char *cgroup_root = "/sys/fs/cgroup", *c;
+        int r;
 
+        (void) mkdir_p(cgroup_root, 0755);
+
+        /* Mount a tmpfs to /sys/fs/cgroup if it's not mounted there yet. */
+        r = path_is_mount_point(cgroup_root, AT_SYMLINK_FOLLOW);
```

Comment:
This causes problems in rkt when cgns is enabled and cgroup-v1 is used.

With cgroup-v1, systemd-nspawn mounts a subtree of the systemd hierarchy RW and the rest RO. In particular, the cpu and memory hierarchies are completely RO.

In rkt, we want to have cpu & memory isolators for some apps running in the nspawn container. For that, we want systemd running in the container to be able to:
- move processes in the correct `system.slice/myapp.service` cgroup in the cpu & memory hierarchies (knob file "cgroup.procs")
- apply some limits (knob file "memory.limit_in_bytes")

To achieve that, rkt pre-mounts the cgroup filesystems and bind-mount a selected set of knob files in read-write mode. The bind mounts done by rkt are not masked by systemd-nspawn because systemd-nspawn takes care of checking with `path_is_mount_point()` if the mount is already done and skip it in that case. This was done on purpose to allow programs calling systemd-nspawn to set-up their own cgroups.

With this PR, we have rkt mounting the cgroups for the container from outside the cgroup namespace. And systemd inside the container still has /proc/sys/fs/cgroups mount with the "outside" view while /proc/$pid/cgroup has the "inside" view.

When /proc/$pid/cgroup and /proc/sys/fs/cgroups works in different cgroup namespaces, systemd in the container cannot work correctly.

As a temporary precaution to avoid a regression, I would like to have an environment variable DISABLE_CGROUPNS=1 or something to allow users to disable that. rkt could set that variable as a temporary solution.

xref https://github.com/coreos/rkt/issues/1757#issuecomment-231760090

/cc @iaguis @poettering